src/mame/misc/babysuprem.cpp

```diff
@@ -33,45 +46,364 @@ class bsuprem_state : public driver_device
+
+	PORT_START("SWITCHES")
+	PORT_BIT(0x01, IP_ACTIVE_LOW, IPT_COIN2)
+	PORT_BIT(0x02, IP_ACTIVE_LOW, IPT_COIN3)
+	PORT_BIT(0x04, IP_ACTIVE_LOW, IPT_CUSTOM) PORT_READ_LINE_DEVICE_MEMBER("hopper_25", ticket_dispenser_device, line_r)
+	PORT_BIT(0x08, IP_ACTIVE_LOW, IPT_CUSTOM) PORT_READ_LINE_DEVICE_MEMBER("hopper_100", ticket_dispenser_device, line_r)
+	PORT_BIT(0x10, IP_ACTIVE_LOW, IPT_COIN1)
+	PORT_BIT(0x20, IP_ACTIVE_LOW, IPT_CUSTOM) PORT_READ_LINE_DEVICE_MEMBER("hopper_5", ticket_dispenser_device, line_r)
+	PORT_BIT(0x40, IP_ACTIVE_LOW, IPT_BUTTON7)
+	PORT_BIT(0x80, IP_ACTIVE_LOW, IPT_BUTTON8)
+
 INPUT_PORTS_END
 
 void bsuprem_state::bsuprem(machine_config &config)
 {
 	// Basic machine hardware
 	Z80(config, m_maincpu, 2.4576_MHz_XTAL); // divider not verified
+	m_maincpu->set_addrmap(AS_PROGRAM, &bsuprem_state::bsuprem_map);
+	m_maincpu->set_periodic_int(FUNC(bsuprem_state::mains_irq), attotime::from_hz(100));
+
+	NVRAM(config, "nvram", nvram_device::DEFAULT_ALL_0);
+
+	HOPPER(config, m_hopper_5, attotime::from_msec(50), TICKET_MOTOR_ACTIVE_HIGH, TICKET_STATUS_ACTIVE_HIGH);
+	HOPPER(config, m_hopper_25, attotime::from_msec(50), TICKET_MOTOR_ACTIVE_HIGH, TICKET_STATUS_ACTIVE_HIGH);
+	HOPPER(config, m_hopper_100, attotime::from_msec(50), TICKET_MOTOR_ACTIVE_HIGH, TICKET_STATUS_ACTIVE_HIGH);
 
 	// Sound hardware
 	SPEAKER(config, "mono").front_center();
 	AY8910(config, m_psg[0], 2.4576_MHz_XTAL); // divider not verified
 	m_psg[0]->add_route(ALL_OUTPUTS, "mono", 0.30);
 	AY8910(config, m_psg[1], 2.4576_MHz_XTAL); // divider not verified
 	m_psg[1]->add_route(ALL_OUTPUTS, "mono", 0.30);
+
+	m_psg[0]->port_a_write_callback().set(FUNC(bsuprem_state::ay8910_u1_porta_w));
+	m_psg[0]->port_b_write_callback().set(FUNC(bsuprem_state::ay8910_u1_portb_w));
+	m_psg[1]->port_a_write_callback().set(FUNC(bsuprem_state::ay8910_u2_porta_w));
+	m_psg[1]->port_b_write_callback().set(FUNC(bsuprem_state::ay8910_u2_portb_w));
+}
+
+void bsuprem_state::bsuprem_map(address_map &map)
+{
+	map(0x0000, 0x0fff).rom();
+	map(0x1000, 0x1fff).rw(FUNC(bsuprem_state::ay8910_w_dsw1_switches2_r), FUNC(bsuprem_state::ay8910_w));
+	map(0x2000, 0x2fff).portr("SWITCHES");
+	map(0x3000, 0x3fff).r(FUNC(bsuprem_state::dsw2_u13_switches_r));
+	map(0x4000, 0x4fff).rom().region("maincpu", 0x2000);
+	map(0x5000, 0x5fff).w(FUNC(bsuprem_state::u9_w));
+	map(0x6000, 0x6fff).w(FUNC(bsuprem_state::u10_w));
+	map(0x7000, 0x77ff).mirror(0x0800).rw(FUNC(bsuprem_state::nvram_r), FUNC(bsuprem_state::nvram_w));
+
+	map(0x8000, 0x8fff).rom().region("maincpu", 0x1000);
+	map(0xc000, 0xcfff).rom().region("maincpu", 0x3000);
+}
+
+void bsuprem_state::machine_reset()
+{
+	m_irq_state = 0;
+	m_u1_porta_data = 0;
+	m_u10_data = 0;
+}
+
+void bsuprem_state::machine_start()
+{
+	m_digits.resolve();
+	m_lamps.resolve();
+	m_buttons.resolve();
+	m_nixie.resolve();
+
+	m_nvram8 = std::make_unique<uint8_t[]>(2048);
+	m_nvram->set_base(m_nvram8.get(),2048);
```

Comment:
Do you need to register this for save states as well?  Also, please put a space after the comma.